### DISCUSSION TITLE
1x1 invertible convolutions logdet

### DISCUSSION THREAD

**Discussion by p0p4k** [coqui-ai/TTS#1799] [General]
While reading [GlowTTS ](https://arxiv.org/abs/2005.11129) and [WaveGlow](https://ieeexplore.ieee.org/document/8683143) papers, they refer to [Glow](https://arxiv.org/abs/1807.03039) (Generative Flow with Invertible 1x1 Convolutions) paper.
In the following Table 1 [Glow paper], we can see the log-determinant (logdet) of the invertible 1x1 convolution (highlighted in yellow). 

![image](https://user-images.githubusercontent.com/8834712/181407572-920d0b0e-b5fc-4600-9e51-56ac4769e2f8.png)

In the 🐸 TTS repo, we use this `logdet` as follows: 
- make det(W) positive, we take the absolute value (*EDITED : thanks @erogol)
https://github.com/coqui-ai/TTS/blob/5094499eba440efd41031ad8d7739c4b49c6045b/TTS/tts/layers/glow_tts/glow.py#L99-L100
- logdet |W| x h x c
https://github.com/coqui-ai/TTS/blob/5094499eba440efd41031ad8d7739c4b49c6045b/TTS/tts/layers/glow_tts/glow.py#L131

I was not sure about the derivation for that formula and no Google search gave me satisfactory answer (especially about the h x w term), so I came up with my own proof as show below. I hope someone who wants to learn the math behind these things can benefit from these.  I would advice at least reading the Glow paper as a prerequisite.

![invconv-1](https://user-images.githubusercontent.com/8834712/178559880-264f927c-3887-4e28-8367-50797370504e.jpg)
![invconv-2](https://user-images.githubusercontent.com/8834712/178559957-638ff72c-7ef3-4628-86ae-87cd60473c63.jpg)
![invconv-3](https://user-images.githubusercontent.com/8834712/178559977-1c0815dc-f122-4f2e-8f49-fea65d98241f.jpg)

> These days, I am exploring the code in 🐸 TTS and comparing with the math in the paper to understand the implementation **line by line**. Doing things like these once a while makes you appreciate the genius behind the deep learning architectures both mathematically and in the form of code. I will post about that as soon as I get some more free time. 
(Previous paper discussions : #1494, #1501, #1506)

**Comment by erogol** [comment]
Thanks for the detailed post. I love the notes :)

It is not actually making W positive, more like making its determinant positive. 

About h and w, they are the height and width of the input. However, in our case, tensors are sequences so we use only length `x_len`. 

I hope it helps with the confusion if there is any.

**Reply by p0p4k** [reply]
Yes, I overlooked the `|det (W)|`  . Thank you for the clarification. Will read VITS and transformer stuff next. The "aha!" moment after understanding the code along with the paper is so addictive!
Although, the 🐸  TTS-papers repo is a more appropriate for posting my notes, I think this repo is much more popular and can increase visibility, hope you don't mind.

**Reply by erogol** [reply]
No worries, post it anywhere you like. Happy to discuss. 

One note about VITS before you start. VITS assumes non-volume changing transformations a.k.a det(w) == 0. Therefore there is no logdet computation there. It is explained in the paper but easy to overlook.
